src/LupaClient.php

```diff
@@ -0,0 +1,191 @@
+<?php
+
+declare(strict_types=1);
+
+namespace LupaSearch;
+
+use GuzzleHttp\Exception\ConnectException;
+use GuzzleHttp\Exception\GuzzleException;
+use GuzzleHttp\Utils;
+use LupaSearch\Exceptions\AuthenticationException;
+use LupaSearch\Exceptions\MissingCredentialsException;
+use LupaSearch\Exceptions\TooManyRetriesException;
+use LupaSearch\Factories\HttpClientFactory;
+use LupaSearch\Factories\HttpClientFactoryInterface;
+use LupaSearch\Utils\JwtUtils;
+use GuzzleHttp\ClientInterface;
+use GuzzleHttp\Exception\ClientException;
+use GuzzleHttp\Psr7\Request;
+
+class LupaClient implements LupaClientInterface
+{
+    /**
+     * @var HttpClientFactoryInterface|null
+     */
+    private $httpClientFactory;
+
+    /**
+     * @var ClientInterface|null
+     */
+    private $httpClient;
+
+    /**
+     * @var string|null
+     */
+    private $jwtToken;
+
+    /**
+     * @var string|null
+     */
+    private $email;
+
+    /**
+     * @var string|null
+     */
+    private $password;
+
+    public function __construct(HttpClientFactoryInterface $httpClientFactory = null)
+    {
+        $this->httpClientFactory = $httpClientFactory ?? new HttpClientFactory();
+    }
+
+    public function getHttpClient(): ClientInterface
+    {
+        if (null === $this->httpClient) {
+            $this->setHttpClient($this->httpClientFactory->create());
+        }
+
+        return $this->httpClient;
+    }
+
+    public function setHttpClient(ClientInterface $httpClient): self
+    {
+        $this->httpClient = $httpClient;
+
+        return $this;
+    }
+
+    public function setJwtToken(?string $jwtToken): self
+    {
+        $this->jwtToken = $jwtToken;
+
+        return $this;
+    }
+
+    public function getJwtToken(): ?string
+    {
+        return $this->jwtToken;
+    }
+
+    public function setEmail(string $email): self
+    {
+        $this->email = $email;
+
+        return $this;
+    }
+
+    public function setPassword(string $password): self
+    {
+        $this->password = $password;
+
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function send(
```

Comment:
I would leave two methods - `sendAuthenticated` and `send`. The first one should call the `send` method, but with headers.